File: java/src/Q124_Binary_Tree_Maximum_Path_Sum.java
import common.TreeNode;

import java.util.Arrays;

public class Q124_Binary_Tree_Maximum_Path_Sum {

    int max = Integer.MIN_VALUE;

    public int maxPathSum(TreeNode root) {
        helper(root);
        return max;
    }

    public int helper(TreeNode node) {
        if (node == null) {
            return 0;
        }

        int left = Math.max(0, helper(node.left));
        int right = Math.max(0, helper(node.right));
        max = Math.max(max, left + right + node.val);
        return Math.max(left, right) + node.val;
    }

    public static void main(String[] args) {
        Q124_Binary_Tree_Maximum_Path_Sum obj = new Q124_Binary_Tree_Maximum_Path_Sum();
        System.out.println(obj.maxPathSum(TreeNode.createTree(new Integer[]{1,2,3})));
        System.out.println(obj.maxPathSum(TreeNode.createTree(new Integer[]{1,-2,3})));
    }

}

File: cpp/Q124_Binary_Tree_Maximum_Path_Sum.cpp
#include <bits/stdc++.h>
#include "common.hpp"
using namespace std;

int maxVal;

int maxPathSum(TreeNode *root)
{
    maxVal = root != 0 ? root->val : 0;
    helper(root);
    return maxVal;
}

int helper(TreeNode *node)
{
    if (node == 0)
    {
        return 0;
    }

    int left = helper(node->left);
    int right = helper(node->right);
    int localMax = max(node->val, max(node->val + left, node->val + right));
    maxVal = max(maxVal, max(localMax, node->val + left + right));
    return localMax;
}

int main()
{
    TreeNode *root = new TreeNode(2, new TreeNode(1), new TreeNode(3));
    cout << maxPathSum(root) << endl;

    return 0;
}